src/core/unit.c

```diff
@@ -1773,6 +1779,11 @@ int unit_start(Unit *u) {
             !unit_test_assert(u))
                 return log_unit_notice_errno(u, SYNTHETIC_ERRNO(EPROTO), "Starting requested but asserts failed.");
 
+        /* Load any custom firewall BPF programs here once to test if they are existing and actually loadable.
+         * Fail here early since later errors in the call chain unit_realize_cgroup to cgroup_context_apply are ignored. */
+        if (unit_get_cgroup_context(u) && bpf_firewall_load_custom(u))
+                return log_unit_notice_errno(u, SYNTHETIC_ERRNO(EPROTO), "Loading custom BPF programs for firewall failed.");
```

Comment:
please do this in unit_prepare_exec() so that this is only done when we execute stuff, not ahead of time.

Reply:
I tried this but it does not work for `systemd-run --scope`, so I would leave it where it is - or any other place?

Reply:
hmm, I think adding it to unit_attach_pids_to_cgroup() might make sense, as that is called in the scope case at the moment we add the first PIDs to the scope. If this is called from both unit_prepare_exec() and unit_attach_pids_to_cgroup() then we'd cover every case where we are about to add PIDs to a cgroup.

Hmm, and  I think it would make sense to make bpf_firewall_load_custom() simply return 0 if unit_get_cgroup_context() doesn't hold. i.e. if a unit doesn't have cgroups, it can't have custom bpf filters either, and we can definitly load "nothing" successfully. That should simplify this a bit.

Also, always propagate the original error. Here bpf_firewall_load_custrom( ) returns a proper error code, but you convert that to EPROTO for no reason...

also, I think bpf_firewall_load_custrom() should log about all errors on its own. that would make the invocation here much simpler, as you don't have to generate a msg here at all..

(background: in systemd we want that functions either log about all errors or about none. In the latter case the caller should log. Hence, whenever you write a function figure out if it shall be a logging or no-logging one, and then stick to it. Logging at LOG_DEBUG level is not considered logging in this case btw, everybody is allowed to do that)